### ISSUE TITLE
/etc/vmware-tools/scripts/suspend-vm-default.d/ not exectuted 

### ISSUE THREAD

**Issue by thelounge-zz** [vmware/open-vm-tools#273]
why? since you stop and start network at suspend/resume there is no valid reason ignore user-scripts and i know nobody which suspends within the guest itself
-------------------
see https://communities.vmware.com/thread/391880

It turns out, for me at least, that the scripts are not run when "Suspend" and "Resume" are used as power options. However, they are run when "Suspend Guest" and "Resume Guest" are used to suspend and resume.
-------------------

**Comment by zhoumgh** [comment]
This is correct. When VMware Tools is not installed/stopped, hangs, or crashes, you can still Suspend/Resume VM. "Suspend Guest" and "Resume Guest" requires VMware Tools running inside guest.

Please refer to:
https://pubs.vmware.com/workstation-9/index.jsp?topic=%2Fcom.vmware.ws.get_started.doc%2FGUID-A4536112-10A6-4574-ADE5-60D9BBAE1F02.html

**Comment by thelounge-zz** [comment]
and how is it exklusive supsend/resume witout vmtools and calling the vmtools-scripts *if* they are installed? i just want suspend/resume calling /etc/vmware-tools/scripts/suspend-vm-default.d/ on the vSphere cluster as you can't guarantee that co-workers are using alway they buttons you wish

some things are really difficult working with VMware (like not can change a damned config option which becomes active on next reboot - who in the world powers down production machines, some actions like EVC change or simply Spectre/Meltown BIOS updates on the host require a power-cylce of the guest instead reboot) with no logical reasoning

**Comment by thelounge-zz** [comment]
the point is that you obviously call thenetwork scripts but NOT /etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh

[root@master:~]$ cat messages
Jul 31 11:47:15 master systemd[1]: Stopping LSB: Bring up/down networking...
Jul 31 11:47:15 master network[972]: Schnittstelle eth0 beenden:  [  OK  ]
Jul 31 11:47:16 master network[972]: Loopback-Schnittstelle beenden:  [  OK  ]
Jul 31 11:47:16 master systemd[1]: Stopped LSB: Bring up/down networking.
Jul 31 11:47:16 master systemd[1]: Starting LSB: Bring up/down networking...
Jul 31 11:47:16 master network[1143]: Loopback-Schnittstelle hochfahren:  [  OK  ]
Jul 31 11:47:16 master kernel: vmxnet3 0000:03:00.0 eth0: intr type 3, mode 0, 2 vectors allocated
Jul 31 11:47:16 master kernel: vmxnet3 0000:03:00.0 eth0: NIC Link is Up 10000 Mbps
Jul 31 11:47:16 master network[1143]: Schnittstelle eth0 hochfahren:  rx unmodified, ignoring
Jul 31 11:47:16 master network[1143]: tx unmodified, ignoring
Jul 31 11:47:16 master network[1143]: no ring parameters changed, aborting
Jul 31 11:47:16 master network[1143]: [  OK  ]
Jul 31 11:47:16 master systemd[1]: Started LSB: Bring up/down networking.

[root@master:~]$ cat /etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh
#!/usr/bin/bash
#########################################################################
# Vmware-Tools-Hook wird vor dem Suspend automatisch aufgerufen         #
# Somit koennen wir Maschinen aus vCenter sauber neustarten             #
#########################################################################

# Aufruf des Hooks ins Syslog schreiben
logger --tag="VMware" "/etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh"

# SIGHUP an ggf. laufendene MySQL-Instanzen um Daten auf Disk zurueckzuschreiben
COUNT_MYSQLD=`ps -C mysqld | grep mysqld | wc -l`
if [ "$COUNT_MYSQLD" -gt "0" ]; then
 logger --tag="VMware" "killall -s SIGHUP mysqld"
 killall -s SIGHUP mysqld
fi

# OS-Buffer auf Disk schreiben
sync

# Sicherstellen dass in jedem Fall Erfolgsstatus zurueckgegeben wird
exit 0

**Comment by zhoumgh** [comment]
You can edit VM settings on ESX UI, go to VM Options | VMware Tools | 'Power Operations', 'Run VMware Tools Scripts', default is to perform soft power operations and run the scripts. The vmx file will have entries like:

toolScripts.afterPowerOn = "TRUE"
toolScripts.afterResume = "TRUE"
toolScripts.beforeSuspend = "TRUE"
toolScripts.beforePowerOff = "TRUE"
...
powerType.powerOff = "soft"
powerType.powerOn = "soft"
powerType.suspend = "soft"
powerType.reset = "soft"

(If you don't see it, default value applies)

Yes, you need to change the VM settings at power off state. Usually people do it at the time VM is created.

The UI buttons act based on the above configurations. In Workstation UI, you can still see menu items 'Suspend Guest' & 'Suspend', 'Suspend Guest' will execute /etc/vmware-tools/suspend-vm-default, while 'Suspend' will not.

**Comment by thelounge-zz** [comment]
"Yes, you need to change the VM settings at power off state" - And for that i have ZERO understanding in 2018! As you can "plan" the virtual Hardware-Upgrade online it's technically possible to just apply the settings and make some yellow warning that there are changes which needs to reboot the virtual machine to take effect and that after ESXi upgrades containing Spectre relevant microcodes you even don't get a hint that you need a *real power cycle* for every guest (besides nobody knows why that as well as EVC mode changes can't be applied due simple reboot at the moment the bios screen appears) is a joke

**Comment by zhoumgh** [comment]
ls -l /etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh

If it is not executable, run

sudo chmod a+x /etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh

Your script will be executed before /etc/vmware-tools/scripts/vmware/network suspend-vm

You can test by just run:

sudo /etc/vmware-tools/suspend-vm-default

Your script should be executed by it.

Clean up with

sudo /etc/vmware-tools/resume-vm-default

**Comment by thelounge-zz** [comment]
i know how to operate Linux systems and besides the Ubuntu world "sudo" is not needed :-)

[root@buildserver:~]$ ls -l /etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh
-rwx------ 1 root root 808 2018-07-22 12:01 /etc/vmware-tools/scripts/suspend-vm-default.d/01-thelounge-basic.sh

**Comment by thelounge-zz** [comment]
> sudo /etc/vmware-tools/suspend-vm-default
> Your script should be executed by it

that damned script should be executed for *both* ways of suspend in case vmware-tools are running, not more and not less is the topic here

**Comment by zhoumgh** [comment]
Correct one thing: on ESX, powerType.suspend default value is "hard", therefore, /etc/vmware-tools/suspend-vm-default will not run.

This is done because we prioritize support use case over other use cases. When guest OS has problem, we want the VM be suspended without changing the guest OS states even if VMware Tools is healthy, so that we can convert the vmss file to guest OS core and analyze it.
